internal/dig/shapes.lua

```diff
@@ -0,0 +1,212 @@
+-- shape definitions for gui/dig
+--@ module = true
+
+if not dfhack_flags.module then
+    qerror("this script cannot be called directly")
+end
+
+-- Base shape class, should not be used directly
+Shape = defclass(Shape)
+Shape.ATTRS {
+    name = "",
+    arr = {},
+    has_point_fn = DEFAULT_NIL,
+    apply_options_fn = DEFAULT_NIL,
+    invert = false,
+    width = 1,
+    height = 1
+}
+function Shape:update(width, height)
+    self.width = width
+    self.height = height
+    self.arr = {}
+    for x = 0, self.width do
+        self.arr[x] = {}
+        for y = 0, self.height do
+            local value = self:has_point_fn(x, y)
+            if not self.invert then
+                self.arr[x][y] = value
+            else
+                self.arr[x][y] = not value
+            end
+        end
+    end
+end
+
+
+-- Shape definitions
+-- All should have a string name, and a function has_point_fn(self, x, y) which returns true or false based
+-- on if the x,y point is within the shape
+-- Also options can be defined in a table, see existing shapes for example
+
+Ellipse = defclass(Ellipse, Shape)
+Ellipse.ATTRS = {
+    name = "Ellipse",
+    has_point_fn = function(self, x, y)
```

Comment:
Would it make sense to define the function in `init()` as I'm going to be also moving the `options` table there? Or at least make it a member function of the class instead of making it a `local function` outside of the class?

Reply:
Ah, yes, good idea. maybe not using an attribute at all and instead declaring it a member function and overriding it OOP-style is the best choice here.

Reply:
As long as that works then I'll do that, I'm not familiar with lua so I wasn't sure exactly what the best approach would be in terms of inheritance so that's probably why I initially put everything in the ATTRS

Reply:
It's not so much a Lua thing as a DFHack widget thing. ATTRS are good for fields that should be set by the parent widget, like the positioning frame and callbacks that link widgets together. Functionally that is not owned by the parent widget might as well just be a regular member function.